### PR TITLE
HPCC-18847 Enable prefetch for alien types and ifblocks

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman ready for initial review.  One change I might need to still make is that alien datatypes still require a prefetch function to be generated (cannot be completely implemented by the type information).  That is also true for datasets with counts/sizes based on other fields.  It might be sensible to fix both of those later at the same time.

File: ecl/hqlcpp/hqlstmt.cpp [hunk 1]
```diff
@@ -1967,6 +1980,9 @@ bool SpecialFunction::queryCombine(const SpecialFunction & next, bool memsetOnly
     {
         if (src != next.src)
             return false;
+        //Only combine skips what have been read from the file, not those that are calculated (e.g., alien types)
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlstmt.cpp]: grammar - "what have been read from the file"

File: rtl/eclrtl/rtlcommon.cpp [hunk 2]
```diff
@@ -142,11 +142,13 @@ const byte * CThorContiguousRowBuffer::peek(size32_t maxSize)
 offset_t CThorContiguousRowBuffer::beginNested()
 {
     size32_t len = readSize();
+    //MORE: If this was ever called from readAhead() then it would need to call noteStartChild();
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlcommon.cpp]: This comment may need more explanation.

File: rtl/eclrtl/rtlcommon.hpp [hunk 3]
```diff
@@ -27,14 +27,17 @@ class ECLRTL_API CThorContiguousRowBuffer : implements IRowDeserializerSource
     virtual size32_t readVStr(ARowBuilder & target, size32_t offset, size32_t fixedSize);
     virtual size32_t readVUni(ARowBuilder & target, size32_t offset, size32_t fixedSize);
 
-    //These shouldn't really be called since this class is meant to be used for a deserialize.
-    //If we allowed padding/alignment fields in the input then the first function would make sense.
+    //The following functions should only really be called when used by the readAhead() function
     virtual void skip(size32_t size);
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlcommon.hpp]: Worth adding overrides ?

File: rtl/eclrtl/rtlcommon.hpp [hunk 4]
```diff
@@ -27,14 +27,17 @@ class ECLRTL_API CThorContiguousRowBuffer : implements IRowDeserializerSource
...
     virtual void skipPackedInt();
     virtual void skipUtf8(size32_t len);
     virtual void skipVStr();
     virtual void skipVUni();
 
+    virtual const byte * querySelf();
```
Comment (richardkchapman) [inline rtl/eclrtl/rtlcommon.hpp]: could perhaps be const?

**Comment by richardkchapman** [comment]
@ghalliday Looks sensible on an initial view other than minor comments and questions. I assume you have checked generated code

File: rtl/eclrtl/rtlcommon.hpp (hunk 4 above)
Comment (ghalliday) [reply rtl/eclrtl/rtlcommon.hpp]: Not sure.  The current implementation might read data if nothing has been read yet.

File: rtl/eclrtl/rtlcommon.hpp (hunk 3 above)
Comment (ghalliday) [reply rtl/eclrtl/rtlcommon.hpp]: Yes - I will modify

**Comment by ghalliday** [comment]
@shamser please review.

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed.  Minor comment. Looks good.

**Pull request merged**